Kyrons branch

KyronMayhew opened the pull request:
Hi, i really like your addon but i found it annoying to manually claim the items and delete the database.
So i updated it it to have an option in the menu bar to get commands to cheat in the items.

When used it generates commands that are under 255 characters ( so that they fit in the Minecraft console window ) that spawn items riding each other, this way each command can spawn around 7 different item types. After all the commands are shown, it offers to clear the database. The UI doesn't automatically refresh, but if you manually refresh it you can see that the claim items are now gone.

Its my first time contributing to someone else's git project so i hope you like my changes enough to pull it into the main branch, feel free to reach out to me at kyronw11@gmail.com if you have any questions about my code.

Commit:
Added button to claim items in the form of commands that spawn several items at once. Also added option to reset the database after claiming the items

Commit:
Recompiled the AnkiCraft.ankiaddon to contain the new addon, and updated the readme usage instructions to reflect the changes

Foxy-null commented:
Thank you so much for the PR! I'm going to review it now!

Foxy-null commented:
Hey @AvianSamurai I've tested your changes in my environment, however it seems like maybe more than 500 items or so are not able to claim.

Command: `/summon item ~ ~ ~ {Item:{id:"experience_bottle",Count:781}, Passengers:[{id:item,Item:{id:"rotten_flesh",Count:168}},{id:item,Item:{id:"oak_log",Count:1037}},{id:item,Item:{id:"sweet_berries",Count:5}},{id:item,Item:{id:"cobblestone",Count:1116}}]}`

What I get:
![image](https://github.com/Foxy-null/AnkiCraft/assets/72304646/e2446b2f-6498-4f53-88c7-eb8441cb9919)

KyronMayhew commented:
Ok, Looks like an issue with integer overflow. I think that Minecraft can only handle counts of up to 255. 
I haven't tested this theory yet but I expect that this is the case for the experience bottle, 781 mod 256 = 13.
Not sure why the rotten flesh didn't spawn. I look into both problems when I'm home and push a fix.

Foxy-null committed:
Update Japanese and remove unnecessary semicolons

Foxy-null committed:
Removal of unnecessary semicolons

Foxy-null commented:
I improved the Google-translated Japanese sentences in `main.py` at [672d47a](https://github.com/Foxy-null/AnkiCraft/pull/17/commits/672d47a565ff5b09f9ebb96eee772820de80b7be). Additionally, I explored ways to fix the bug.

My suggestion is to split the count and repeat the process if the number of one type of item exceeds 255. I'll let you know if I can make the changes to the code satisfactorily.

KyronMayhew commented:
Ok so the problem is even weirder than i thought. Looks like Minecraft will only summon up to 127 items. Trying to summon more than that summons nothing, unless you try to summon more than 256 items, then it will integer overflow. 

I'm working on fixing it now, my plan is to generate multiple stacks of the same item in the summon command so it can still be grouped with other items to reduce the amount of commands needed. If it doesn't all fit into a summon command, then switch to a give command that can give up to 6400 items. If there are more than 6400 items in a single item type, multiple give commands will have to be run

Commit:
Fixed invalid item number issues, potions, fireworks and version support

Commit:
minor change to naming of config line

KyronMayhew commented:
@Foxy-null I fixed the issues with players being given invalid numbers of items.

I also found that my code for giving the players fireworks and the healing potion wasn't working correctly as these items require NBT data. unfortunately Mojang changed the formatting for this data a month ago in version 1.20.5 so I added support for both versions in the config.

When fireworks are given, since it needs NBT data and I couldn't decide on a single nice firework design, I built it to randomize the data, giving different fireworks every time.

Please let me know if you find any bugs, I did a fair bit of testing and fixed what I found. Happy to fix anything bugs you find.

Foxy-null commented:
Thanks for the really quick fix! I'm going to take a look now!

Foxy-null committed:
Automatically copy to clipboard feature

Foxy-null commented:
Hey, still some items are wrongly shown :(

  What about providing all items in the form of Give commands, so they can be used equally by Bedrock edition players as well as Java edition?
  It would also be easy to customize the username of the user to whom the item will be given in the config.json.

  BTW, I've added a feature to automatically copy the command to the clipboard when any window appears.
![image](https://github.com/Foxy-null/AnkiCraft/assets/72304646/1aee6fb7-2059-4863-a90d-7311b2a259ff)

Foxy-null committed:
Add pyperclip module

Foxy-null commented:
Just in case, I'll share my AnkiCraft data file with you.

## [Here's the link to download my save data](https://d.kuku.lu/g86b55hcz)

KyronMayhew commented:
Thanks for providing your database file, that will help a lot. I will definitely investigate further. Also, I am happy to change the commands over to all be /give for the bedrock edition. My once concern NBT data is handled differently in bedrock than in Java. so for the health potion and fireworks to work (even without the firework randomization) ill probably need the user to define whether they are using bedrock edition or not in the config.

I'm going to be a bit busy with work, so I'll probably push that fix in the next day or two

Commit:
changed to use /give command and supported bedrock

Commit:
Merge remote-tracking branch 'origin/KyronsBranch' into KyronsBranch

Commit:
Duplicate items now get summed together before commands are shown to reduce amount of commands

Commit:
removed (java edition only) from message box

KyronMayhew commented:
Alright, i ended up just looking at it now. Changing to claiming items one at a time with /give greatly reduced code complexity, but it increased the number of commands needed to 85. But i noticed a lot of these items where duplicates between different biomes, So i changed it to sum the count of duplicate items from different biomes together first, reducing the count to 73 commands. 

I also added a target player option to the config that defaults to "@p" but can be changed to any Minecraft username. I also added a config option for players using bedrock edition, all this changes is the commands for the health potion and firework. Unfortunately, you cant specify NBT data in bedrock so the player does not receive random fireworks in bedrock edition.

I tested with your database, I ran all 85 commands and crossed them off the list to make sure each worked and generated the correct items. Then after i made the summing modification, i tested several of the commands to make sure the summed counts were right. 

Don't rush to review my changes, you've already reviewed them 3 times and i feel bad for taking up any more of your time. As always let me know if you find any issues or have any ideas for modifications

Foxy-null commented on addon/config.py:
```diff
@@ -73,6 +73,8 @@ def _writeAddonMeta(meta):
 
         """
 
+        unicode = str # Compatibility with python 3.9
```
Thanks!

Foxy-null commented on addon/menu.py:
```diff
@@ -159,3 +180,144 @@ def check_correct_game_in_menu(menu_actions_by_game_id, load_current_game_id):
+        
+        return f'/give {username} firework_rocket{{Fireworks:{{Flight:{flight_time},Explosions:[{{Type:{shape},Flicker:{has_twinkle}b,Trail:{has_trail}b,Colors:[I{colour}],FadeColors:[I{fade_colour}]}}]}}}} {cmd_count}', final_count
+
+        # I cannot stress how annoyed i was to find out the minecraft changed the way you give items with nbt data a month ago
```
I didn't know that there was a change to the NBT system, and I'm sorry to hear that...

Foxy-null commented on addon/streaks.py:
```diff
@@ -410,15 +412,17 @@ def medal_sound_src(self):
             medal_image=image_path("halo_5/minecraft-bottle-o-enchanting.gif"),
             medal_sound=sound_path("orb"),
-            name="Bottle o' Enchant ing",
+            name="Bottle o' Enchanting",
```
Thx!

Foxy-null commented on addon/streaks.py:
```diff
@@ -925,7 +966,8 @@ def medal_sound_src(self):
         name_jp="革",
         game_id="halo_5",
-        rank=5,
+        rank=7,
```
Oh! I didn't notice this mistake! Thank you for checking my code carefully.

Foxy-null committed:
Touch-up and add some stuff :)

Update config.md

Foxy-null committed:
Update AnkiCraft.ankiaddon

KyronMayhew commented on addon/streaks.py:
oh my god you added the pufferfish 😂
Love your work

Foxy-null replied:
Oh no, you've already noticed that! Too fast🤣

Foxy-null commented:
> Don't rush to review my changes, you've already reviewed them 3 times and i feel bad for taking up any more of your time.

Don't feel bad, I'm really glad that someone cares about this add-on. It is only natural that I should express my utmost respect to you for putting so much passion into this add-on!

What?!?!
I didn't know that leaving a message on a pull request would send notifications here!
Sorry for spamming your notification!

By the way, I've finished reviewing your code and I'm going to make an update now!
I really appreciate your work! You're the best!
Thanks a lot again!🩵
